File: Code/CSharp/Adapter/adapter.cs
public interface ITrain {
    void setLine(int lineNumber);
    void acceptPassengers();
    void stop();
    void checkTrainTicket(string code);
    void drive();
}

public class Train : ITrain {
    private int wagonCount;
    protected string line;
    protected int passengerCount;

    public Train(int wagonCount) {
        this.wagonCount = wagonCount;
    }

    public virtual void setLine(int lineNumber) {
        line = "T" + lineNumber;
    }

    public virtual void acceptPassengers(){
        passengerCount = wagonCount*50;
    }

    public virtual void stop() {
        passengerCount -= wagonCount*50;
    }

    public void checkTrainTicket(string code) {
        if (code != "train") {
            System.Console.WriteLine("Error");
        } 
    }
    public virtual void drive() {
        System.Console.WriteLine("Train " + line + " driving with " + passengerCount + " passengers");
    }
}

public class Bus {
    protected int capacity;
    protected string line;
    protected int passengerCount;

    public Bus(int capacity) {
        this.capacity = capacity;
    }

    public virtual void setLine(int lineNumber) {
        line = "B" + lineNumber;
    }

    public void acceptPassengers() {
        passengerCount = capacity;
    }

    public void releasePassengers() {
        passengerCount = 0;
    }

    public void checkBusTicket(string code) {
        if (code != "bus") {
            System.Console.WriteLine("Error");
        }
    }

    public void drive() {
        System.Console.WriteLine("Bus " + line + " driving with " + passengerCount + " passengers");
    }
}

public class RailroadController {
    public static void runLine(ITrain train, int n) {
        train.setLine(n);
        train.acceptPassengers();
        train.checkTrainTicket("train");
        train.drive();
        train.stop();
    }
}
public class AccessibleTrain : Train {
    public AccessibleTrain(int wagonCount) : base(wagonCount){
    }

    public string Line {
        get {
            return line;
        }
    }
}

public class TrainReplacement : Bus, ITrain {
    AccessibleTrain train;
    public TrainReplacement(int capacity) : base(capacity) {
        train = new AccessibleTrain(1);
    }
    public void stop() {
        this.releasePassengers();
    }

    public void checkTrainTicket(string code) {
        train.checkTrainTicket(code);
    }

    public override void setLine(int lineNumber) {
        train.setLine(lineNumber);
        line = train.Line;
    }
}

public class Tester {
    public static void Main(string[] args) {
        ITrain replacementTrain = new TrainReplacement(40);
        RailroadController.runLine(replacementTrain, 2);
    }
}

File: Code/Go/Adapter/adapter.go
package main

import (
	"fmt"
	"strconv"
)

type railroadRunner interface {
	setLine(int)
	acceptPassengers()
	stop()
	drive()
	checkTrainTicket(string)
}

type Train struct {
	wagonCount     int
	line           string
	passengerCount int
}

type Bus struct {
	capacity       int
	line           string
	passengerCount int
}

type TrainReplacement struct {
	*Bus
	train *Train
}

type RailroadController struct {
}

func NewTrain(wagonCount int) *Train {
	return &Train{wagonCount, "", 0}
}

func (train *Train) setLine(lineNumber int) {
	train.line = "T" + strconv.Itoa(lineNumber)
}

func (train *Train) acceptPassengers() {
	train.passengerCount = 50 * train.wagonCount
}

func (train *Train) stop() {
	train.passengerCount -= 50 * train.wagonCount
}

func (train *Train) drive() {
	fmt.Println("Train " + train.line + " drives with " + strconv.Itoa(train.passengerCount) + " passengers")
}

func (train *Train) checkTrainTicket(code string) {
	if code != "train" {
		fmt.Println("Error")
	}
}

func NewBus(capacity int) *Bus {
	return &Bus{capacity, "", 0}
}

func (bus *Bus) setLine(lineNumber int) {
	bus.line = "B" + strconv.Itoa(lineNumber)
}

func (bus *Bus) acceptPassengers() {
	bus.passengerCount = bus.capacity
}

func (bus *Bus) releasePassengers() {
	bus.passengerCount = 0
}

func (bus *Bus) drive() {
	fmt.Println("Bus " + bus.line + " drives with " + strconv.Itoa(bus.passengerCount) + " passengers")
}

func (bus *Bus) checkBusTicket(code string) {
	if code != "bus" {
		fmt.Println("Error")
	}
}

func NewTrainReplacement(capacity int) TrainReplacement {
	return TrainReplacement{NewBus(capacity), NewTrain(1)}
}

func (rTrain TrainReplacement) checkTrainTicket(code string) {
	rTrain.train.checkTrainTicket(code)
}

func (rTrain TrainReplacement) stop() {
	rTrain.releasePassengers()
}

func (rTrain TrainReplacement) setLine(lineNumber int) {
	rTrain.train.setLine(lineNumber)
	rTrain.line = rTrain.train.line
}

func (rc RailroadController) runLine(train railroadRunner, lineNumber int) {
	train.setLine(lineNumber)
	train.acceptPassengers()
	train.checkTrainTicket("train")
	train.drive()
	train.stop()
}

func main() {
	train := NewTrainReplacement(40)
	RailroadController{}.runLine(train, 2)
}
